### PR TITLE
Sparse: add more warnings to detect more errors

### PR DESCRIPTION
**Pull request by plbossart**
'man sparse' gives a long set of options, I chose to enable them all except for 4 options that throw errors that aren't real or useful.

Also I added compilation support for other Intel common/ and boards/ which can be problematic at times, and drivers/soundwire/

Feedback welcome.

### CODE REVIEW

**Comment by plbossart** [comment]
Wow. Sparse found a single character bug in the SOF/SoundWire machine driver selection, and of course in code that I 'optimized'. Boy am I glad I enabled more warnings :-)

@RanderWang @bardliao please double-check the last patch ('ASoC: SOF: Intel: hda: fix index used in inner loop'), I don't know what I was thinking and how we tested this but it looks absolutely wrong.

File: .github/workflows/buildtest.yml [hunk 1]
```diff
@@ -33,10 +33,13 @@ jobs:
...
         run: |
-          export ARCH=x86_64 CF="-Wsparse-error"
+          export ARCH=x86_64 CF="-Wsparse-error -Wsparse-all -Wno-bitwise-pointer -Wno-pointer-arith -Wno-typesign -Wnoshadow"
           bash kconfig/kconfig-sof-default.sh
           make modules_prepare
-          make -k M=sound/soc/sof C=2
+          make -k sound/soc/sof/ C=2
+          make -k sound/soc/intel/common/ C=2
+          make -k sound/soc/intel/boards/ C=2
+          make -k drivers/soundwire/ C=2
```
Comment (marc-hb) [inline .github/workflows/buildtest.yml:42]: If there are warnings in `sound/soc/sof/` then I'm 97% sure the others won't run. Is that what you want? If this is not what you want I see two possible solutions:

1. Separate github jobs. Pros: faster completion and smaller logs in each job. Cons: a lot of duplication in this `.yml` file (github does not support YAML anchors

2. `make -k sound/soc/sof/  sound/soc/intel/common ... C=2`

File: .github/workflows/buildtest.yml [hunk 2]
```diff
@@ -33,10 +33,13 @@ jobs:
 
       - name: build start
```
Comment (marc-hb) [inline .github/workflows/buildtest.yml:34]: Can you rename this to something less generic like "run sparse"? It is the main section after all.

File: .github/workflows/buildtest.yml (hunk 1 above)
Comment (plbossart) [reply .github/workflows/buildtest.yml:42]: this is what we do for other cases, I don't think it makes sense to continue compiling a different directory if the first one is already broken. The -k is already questionable, but I guess it's useful for people to correct all the errors in that directory.

File: .github/workflows/buildtest.yml (hunk 2 above)
Comment (plbossart) [reply .github/workflows/buildtest.yml:34]: all the file has 'build start' for each job, I have no idea why but that would need to be a separate change.

File: .github/workflows/buildtest.yml (hunk 1 above)
Comment (marc-hb) [reply .github/workflows/buildtest.yml:42]: Sounds like a good trade-off, just wanted to make sure you were aware of this.

File: sound/soc/sof/intel/hda.c [hunk 3]
```diff
@@ -1117,7 +1117,7 @@ static bool link_slaves_found(struct snd_sof_dev *sdev,
 
 			/* find out how many identical parts are expected */
 			for (k = 0; k < link->num_adr; k++) {
-				u64 adr2 = link->adr_d[i].adr;
+				u64 adr2 = link->adr_d[k].adr;
```
Comment (bardliao) [inline sound/soc/sof/intel/hda.c:1120]: Nice catch. The reason how it worked is that we only support the same codecs on a single link. i.e. adr_d[0] == adr_d[1] == adr_d[2] etc for mfg_id, part_id and link_id. So we can still count `expected_part_count` using the same adr_d[i].adr.

**Review by paulstelian97 — APPROVED** [review]

**Review by lyakh — APPROVED** [review]

File: sound/soc/sof/intel/hda.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda.c:1120]: @plbossart Maybe above should be added to the commit message.

**Review by kv2019i — COMMENTED** [review]
I'm not super fond of the "ull" postfixes, but given we have so few sparse warnings otherwise and we can get to a clean run, probably worth it. I'd change the last commit msg to something a bit less alarming (now that Bard provided an explanation).

**Comment by plbossart** [comment]
> I'm not super fond of the "ull" postfixes, but given we have so few sparse warnings otherwise and we can get to a clean run, probably worth it. I'd change the last commit msg to something a bit less alarming (now that Bard provided an explanation).

I'll fix the commit message before upstream, there are already multiple approvers.

**Pull request merged**